File: solutions/2792. Count Nodes That Are Great Enough/2792.py
class Solution:
  def countGreatEnoughNodes(self, root: TreeNode | None, k: int) -> int:
    ans = 0

    def dfs(root: TreeNode | None) -> list[int]:
      nonlocal ans
      if not root:
        return []

      kSmallest = sorted(dfs(root.left) + dfs(root.right))[:k]
      if len(kSmallest) == k and root.val > kSmallest[-1]:
        ans += 1

      return kSmallest + [root.val]

    dfs(root)
    return ans

File: solutions/2792. Count Nodes That Are Great Enough/2792.java
class Solution {
  public int countGreatEnoughNodes(TreeNode root, int k) {
    dfs(root, k);
    return ans;
  }

  private int ans = 0;

  private Queue<Integer> dfs(TreeNode root, int k) {
    if (root == null)
      return new PriorityQueue<>(Collections.reverseOrder());

    Queue<Integer> kSmallest = dfs(root.left, k);
    Queue<Integer> kSmallestRight = dfs(root.right, k);
    kSmallest.addAll(kSmallestRight);

    while (kSmallest.size() > k)
      kSmallest.poll();
    if (kSmallest.size() == k && root.val > kSmallest.peek())
      ++ans;

    kSmallest.offer(root.val);
    return kSmallest;
  }
}
